var/spack/repos/builtin/packages/megadock/package.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+import os
+
+
+class Megadock(MakefilePackage, CudaPackage):
+    """an ultra-high-performance protein-protein docking for
+       heterogeneous supercomputers"""
+
+    homepage = "http://www.bi.cs.titech.ac.jp/megadock/"
+    url      = "http://www.bi.cs.titech.ac.jp/megadock/archives/megadock-4.0.3.tgz"
+
+    version('4.0.3', sha256='c1409a411555f4f7b4eeeda81caf622d8a28259a599ea1d2181069c55f257664')
+
+    variant('mpi', description='Enable MPI', default=False)
+    variant('cuda', description='Enable CUDA', default=False)
+
+    depends_on('fftw')
+
+    depends_on('mpi', when='+mpi')
+    depends_on('cuda', when='+cuda')
+
+    def edit(self, spec, prefix):
+        # point cuda samples to cuda prefix
+        filter_file('/opt/cuda/6.5/samples', '$(CUDA_INSTALL_PATH)/samples',
+                    'Makefile', string=True)
+
+        # sneak link to -lm for sin(), cos()
+        filter_file('-o calcrg', '-lm -o calcrg', 'Makefile', string=True)
+
+        # don't force a cuda architecture
+        filter_file('-arch=$(SM_VERSIONS) ', '', 'Makefile', string=True)
+
+        # set mpi, cuda installations
+        if '+mpi' in spec:
+            env['MPICOMPILER'] = self.spec['mpi'].mpicxx
+
+        if '+cuda' in spec:
+            env['CUDA_INSTALL_PATH'] = self.spec['cuda'].prefix
+
+        env['FFTW_INSTALL_PATH'] = self.spec['fftw'].prefix
+        env['CPPCOMPILER'] = 'c++'
```

Comment:
Setting environment variables should be done in the `setup_environment` method. For cases like this one you should modify `spack_env` (the first argument to the function). As an example implementation:

https://github.com/spack/spack/blob/d4d402586c949b167334c7f53f2889017ae6ce9d/var/spack/repos/builtin/packages/go/package.py#L97-L102